### ISSUE TITLE
Error serialization must handle non-errors better

### ISSUE THREAD

**Issue by novemberborn** [avajs/ava#1562]
[`serialize-error`](https://github.com/avajs/ava/blob/c72f4f24bc68b1e0595ff83597e9072d8c886949/lib/serialize-error.js#L49) currently does not guard against being called with a `null` or `undefined` value.

```js
import test from 'ava'

test.cb(t => {
  setImmediate(() => {throw null})
  setImmediate(t.end)
})
```

results in:

```console
❯ npx ava

  1 passed
  1 exception

  Uncaught Exception
  Error: Failed to serialize uncaught exception
    at process.on.exception (/private/var/folders/2_/qczp184x76b2nl034sq5hvxw0000gn/T/tmp.dcvW5BUyA5/node_modules/ava/lib/test-worker.js:93:15)
    at emitOne (events.js:115:13)
    at process.emit (events.js:210:7)
    at process._fatalException (bootstrap_node.js:374:26)
```

Similarly, if a primitive is thrown, the output is terrible:

```js
import test from 'ava'

test.cb(t => {
  setImmediate(() => {throw 5})
  setImmediate(t.end)
})
```

```console
❯ npx ava

  1 passed
  1 exception

  Uncaught Exception
  Threw non-error: {"avaAssertionError":false,"source":null,"object":5,"type":"exception","file":"test.js"}
```

The same goes when throwing objects:

```js
import test from 'ava'

test.cb(t => {
  setImmediate(() => {throw new Date()})
  setImmediate(t.end)
})
```

```console
❯ npx ava

  1 passed
  1 exception

  Uncaught Exception
  Threw non-error: {"avaAssertionError":false,"source":null,"object":{},"type":"exception","file":"test.js"}
```

If the error doesn't seem to be an error we should [format it](https://github.com/avajs/ava/blob/c72f4f24bc68b1e0595ff83597e9072d8c886949/lib/assert.js#L20-L29) and print that instead.

**Comment by sindresorhus** [comment]
WIth:

```js
throw 1;
```

I would print:

```
Threw non-error: 1
YOU SHOULD NOT THROW NON-ERRORS. GO THINK ABOUT WHAT YOU DID.
```

Jokes aside, I'm serious about including the first sentence in some form. People should learn. The fact that it's even allowed in JS is very JS, but also very annoying.

**Comment by novemberborn** [comment]
Here we're dealing with errors that are not handled in the test or by assertions. They could even be bugs (`throw err` where `err` is an undefined variable). I don't think we need to proselytize in the uncaught exception handler. That's what `t.throws()` is for.

**Comment by sindresorhus** [comment]
Ok

**Comment by novemberborn** [comment]
This was fixed in #1722.
